### PR TITLE
ISSUE-4471 integrate cluster select query with new processor

### LINKED ISSUE #4471: integrate cluster select query with new processor

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Refine query fragment code
- integrate cluster with new processor
- Support push base model data exchange
- Preparing to remove the flight API(initial stage)
- Resource control (multiple query fragments will share the same resource(cpu and network) and refer to max_threads)
- Support multiple types of data transfer between cluster nodes(e.g., logs, metrics, progress, commit_entity, initial stage only support progress)

## Changelog

- New Feature

## Related Issues

Fixes #4471

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by xudong963** [comment]
Please ping me when the ticket is ready, I want to help review it!

**Comment by zhang2014** [comment]
> Please ping me when the ticket is ready, I want to help review it!

OK, I'll add you to the reviewer

**Comment by fkuner** [comment]
Is there any RFC or doc about this pr?  For learning conveniently.

File: query/src/api/rpc/exchange/data_exchange.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    pub fn create(destination_ids: Vec<String>, exchange_expression: Expression) -> DataExchange {
+        DataExchange::HashDataExchange(HashDataExchange {
+            destination_ids,
+            exchange_expression,
+        })
+    }
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct MergeExchange {
```
Comment (leiysky) [inline query/src/api/rpc/exchange/data_exchange.rs:34]: Shall we support different merge mode in the future? For example sorted-merge, concat merge, random merge etc.

File: query/src/api/rpc/exchange/exchange_channel.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+use async_channel::Sender;
+use common_arrow::arrow_format::flight::data::FlightData;
+use common_exception::Result;
+
+use crate::api::rpc::packet::DataPacket;
+
+// Different from async_channel::Sender
+// It is allowed to close the channel when has one reference.
+// In other words, we only record the number of cloned.
+pub struct FragmentSender {
```
Comment (leiysky) [inline query/src/api/rpc/exchange/exchange_channel.rs:16]: Does the term `Fragment` means a **data chunk** to transport?

Comment (zhang2014) [reply query/src/api/rpc/exchange/exchange_channel.rs:16]: It's Query Fragment sender.

File: query/src/api/rpc/exchange/data_exchange.rs (hunk 1 above)
Comment (zhang2014) [reply query/src/api/rpc/exchange/data_exchange.rs:34]: Yes, if necessary.

File: query/src/api/rpc/exchange/data_exchange.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_planners::Expression;
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub enum DataExchange {
+    // None,
+    Merge(MergeExchange),
+    ShuffleDataExchange(ShuffleDataExchange),
```
Comment (leiysky) [inline query/src/api/rpc/exchange/data_exchange.rs:21]: Does that mean we haven't support broadcast exchange in current PR yet?

**Review by leiysky — APPROVED** [review]
Rest LGTM

**Comment by Xuanwo** [comment]
Build failed, triggered rerun.

**Comment by Xuanwo** [comment]
> @zhang2014 , Do we introduce the `rwlock` again? I test this PR, and the performance has a big drop.
> 
> ```
> mysql> SELECT avg(number) FROM numbers_mt(100000000000);
> +------------------+
> | avg(number)      |
> +------------------+
> | 9323559.74711512 |
> +------------------+
> 1 row in set (19.78 sec)
> ```

As this PR has been merged, I converted this comment into a new issue.

File: query/src/api/rpc/exchange/data_exchange.rs (hunk 3 above)
Comment (zhang2014) [reply query/src/api/rpc/exchange/data_exchange.rs:21]: Yes, this requires some work, and I will complete it in the next pr.

**Pull request merged**